crates/sui-framework/sources/collectible/kiosk.move

```diff
@@ -0,0 +1,564 @@
+/// which gives them a power to enforce any constraint on trades by
+/// either using one of the pre-built primitives (see `sui::royalty`)
+/// or by implementing a custom policy. The latter requires additional
+/// support for discoverability in the ecosystem and should be performed
+/// within the scope of an Application or some platform.
+///
+/// - A type can not be traded in the Kiosk unless there's a policy for it.
+/// - 0-royalty policy is just as easy as "freezing" the `AllowTransferCap`
+///   making it available for everyone to authorize deals "for free"
+///
+/// 2. Traders: anyone can create a Kiosk and depending on whether it's
+/// a shared object or some shared-wrapper the owner can trade any type
+/// that has issued `TransferPolicyCap` in a Kiosk. To do so, they need
+/// to make an offer, and any party can purchase the item for the amount of
+/// SUI set in the offer. The responsibility to follow the transfer policy
+/// set by the creator of the `T` is on the buyer.
+///
+/// 3. Marketplaces: marketplaces can either watch for the offers made in
+/// personal Kiosks or even integrate the Kiosk primitive and build on top
+/// of it. In the custom logic scenario, the `TransferPolicyCap` can also
+/// be used to implement application-specific transfer rules.
+///
+module sui::kiosk {
+    use std::option::{Self, Option};
+    use sui::object::{Self, UID, ID};
+    use sui::dynamic_object_field as dof;
+    use sui::dynamic_field as df;
+    use sui::transfer_policy::{Self, TransferRequest};
+    use sui::tx_context::{TxContext, sender};
+    use sui::balance::{Self, Balance};
+    use sui::coin::{Self, Coin};
+    use sui::sui::SUI;
+    use sui::event;
+
+    // Collectible is a special case to avoid storing `Publisher`.
+    friend sui::collectible;
+
+    /// Trying to withdraw profits and sender is not owner.
+    const ENotOwner: u64 = 0;
+    /// Coin paid does not match the offer price.
+    const EIncorrectAmount: u64 = 1;
+    /// Trying to withdraw higher amount than stored.
+    const ENotEnough: u64 = 2;
+    /// Trying to close a Kiosk and it has items in it.
+    const ENotEmpty: u64 = 3;
+    /// Attempt to take an item that has a `PurchaseCap` issued.
+    const EListedExclusively: u64 = 4;
+    /// `PurchaseCap` does not match the `Kiosk`.
+    const EWrongKiosk: u64 = 5;
+    /// Tryng to exclusively list an already listed item.
+    const EAlreadyListed: u64 = 6;
+    /// Trying to call `uid_mut` when extensions disabled
+    const EExtensionsDisabled: u64 = 7;
+
+    /// An object which allows selling collectibles within "kiosk" ecosystem.
+    /// By default gives the functionality to list an item openly - for anyone
+    /// to purchase providing the guarantees for creators that every transfer
+    /// needs to be approved via the `TransferPolicy`.
+    struct Kiosk has key, store {
+        id: UID,
+        /// Balance of the Kiosk - all profits from sales go here.
+        profits: Balance<SUI>,
```

Comment:
Is open to all token types considered？，The difficulty seems to have increased, but it's more versatile

Reply:
Currently it only allows SUI. It's the main use case and not having a second type parameter simplifies interaction a bunch.